### PR TITLE
Interleaved MHA for CPU path

### PR DESCRIPTION
**Pull request by TaoLv**
## Description ##
This PR fills the CPU counterpart of PR https://github.com/apache/incubator-mxnet/pull/16408.
Only FP32 is supported yet.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by TaoLv** [comment]
@eric-haibin-lin @pengzhao-intel @Caenorst Please help to review. I copied the UTs from test_operator_gpu.py to test_operator.py but didn't remove the original ones since it seems they have minimal cuda version requirement. Let me know if it's not appropriate. Thanks!

File: tests/python/unittest/test_operator.py [hunk 1]
```diff
@@ -9373,6 +9373,330 @@ def check_random_uniform():
...
+    assert_allclose(att_score_orig, att_score_opti, rtol=1e-2, atol=1e-3)
+    assert_allclose(output_orig, output_opti, rtol=1e-2, atol=1e-3)
+
+    for k in grads_opti.keys():
+        assert(grads_orig[k].dtype == grads_opti[k].dtype)
+        assert(grads_orig[k].shape == grads_opti[k].shape)
+        assert_allclose(grads_orig[k], grads_opti[k], rtol=1e-2, atol=1e-3)
+
+@with_seed(12345)
+def test_multihead_attention_encdec():
```
Comment (eric-haibin-lin) [inline tests/python/unittest/test_operator.py]: do we still need `check_multihead_attention_selfatt ` in test_operator_gpu.py?

File: src/operator/contrib/transformer.cc [hunk 2]
```diff
@@ -122,6 +122,531 @@ static bool InterleavedMatMulEncDecValAttShape(const NodeAttrs& attrs,
   return true;
 }
 
+void strided_batch_sgemm(bool transA, bool transB,
```
Comment (eric-haibin-lin) [inline src/operator/contrib/transformer.cc]: Is it better to move this function to the file with linalg prefix?

File: src/operator/contrib/transformer.cc [hunk 3]
```diff
@@ -122,6 +122,531 @@ static bool InterleavedMatMulEncDecValAttShape(const NodeAttrs& attrs,
...
+                        kv_seq_len,
+                        kv_seq_len * q_seq_len,
+                        beta,
+                        queries_grads,
+                        lead_dim_q,
+                        batch_stride_q,
+                        attn_batches);
+  }
+  if (req[1] != kNullOp) {
+    if (req[1] == kWriteTo) {
```
Comment (eric-haibin-lin) [inline src/operator/contrib/transformer.cc]: Shall we also check req[1] == kWriteInplace?

File: src/operator/contrib/transformer.cc [hunk 4]
```diff
@@ -122,6 +122,531 @@ static bool InterleavedMatMulEncDecValAttShape(const NodeAttrs& attrs,
...
+  const index_t attn_batches      = inputs[2].shape_[0];
+  const index_t q_seq_len         = inputs[2].shape_[1];
+  const index_t embed_dim         = output_lin_kv_dim / 2;
+  const index_t head_dim          = embed_dim / params.heads;
+  const index_t lead_dim_kv       = attn_batches * head_dim * 2;
+  const index_t batch_stride_kv   = 2 * head_dim;
+  const float alpha               = 1.f;
+
+  if (req[0] != kNullOp) {
+    if (req[0] == kWriteTo) {
```
Comment (eric-haibin-lin) [inline src/operator/contrib/transformer.cc]: Shall we also check req[1] == kWriteInplace?

File: src/operator/contrib/transformer.cc (hunk 3 above)
Comment (eric-haibin-lin) [reply src/operator/contrib/transformer.cc]: Shall we also document why memset is required here

**Comment by TaoLv** [comment]
> could you elaborate "it seems they have minimal cuda version requirement" and why we duplicate the test?

Because of this line: https://github.com/apache/incubator-mxnet/blob/master/tests/python/gpu/test_operator_gpu.py#L2863
But since the oldest cuda version in CI is 9.2 so I think it's safe to remove this check and copy tests to test_operator.py. I will remove the duplicated part in test_operator_gpu.py.

**Comment by eric-haibin-lin** [comment]
Thanks for the reply. Look forward to the update

File: src/operator/contrib/transformer.cc (hunk 2 above)
Comment (TaoLv) [reply src/operator/contrib/transformer.cc]: I don't expect this function will be widely used except these mha operators. Also linalg* functions need high level abstraction but I'm not sure how to wrap the pointers a/b/c here into Tensor while still keeping the m/n/k and strides arguments, and how to implement the gpu part.

File: src/operator/contrib/transformer.cc (hunk 3 above)
Comment (TaoLv) [reply src/operator/contrib/transformer.cc]: > Shall we also check req[1] == kWriteInplace?

FInplaceOption is not registered in this backward operator. Do we still need check kWriteInplace here?

Comment (TaoLv) [reply src/operator/contrib/transformer.cc]: > Shall we also document why memset is required here

Reference to the implementation on gpu path: https://github.com/apache/incubator-mxnet/blob/master/src/operator/contrib/transformer.cu#L160.
I thought it relates to the discussion about the bwd_ignore_zero_init flag and gradient accumulation in the original PR. @Caenorst any comments here?

Comment (eric-haibin-lin) [reply src/operator/contrib/transformer.cc]: @Caenorst shall we change it to `req[1] == kWriteTo || req[1] == kWriteInplace ` for GPU code path, too?

**Comment by pengzhao-intel** [comment]
Is this PR good to merge? @TaoLv @eric-haibin-lin

**Comment by TaoLv** [comment]
@eric-haibin-lin Code is re-based. `memset` is needed to pass the unit tests. We can revisit the kWriteInplace check in a follow up PR if we notice any performance problem there. As the FInplaceOption is not enabled for these operators, I think it's safe for not checking it. Let me know what do you think. Thnanks!

**Comment by szha** [comment]
>  But since the oldest cuda version in CI is 9.2 so I think it's safe to remove this check and copy tests to test_operator.py. I will remove the duplicated part in test_operator_gpu.py.

@TaoLv  This is not true for the CD pipeline for cu90. See [this failure](http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/restricted-mxnet-cd%2Fmxnet-cd-release-job/detail/mxnet-cd-release-job/494/pipeline/647#step-1029-log-1633)

**Comment by TaoLv** [comment]
@szha Any suggestion for this case? It's shared for both CPU and GPU test, can we add the cuda requirement decorator to it? Also if we're still releasing cu90 flavor, why it was removed from CI?

**Pull request merged**